src/test/pthreads/test_pthreads.rs

```diff
@@ -0,0 +1,489 @@
+/*
+ * The Shadow Simulator
+ * See LICENSE for licensing information
+ */
+
+use std::error::Error;
+
+use test_utils::set;
+use test_utils::TestEnvironment as TestEnv;
+
+const NUM_THREADS: usize = 5;
+
+const SUCCESS: isize = 0;
+const MUTEX_LOCK_FAILED: isize = -1;
+const MUTEX_UNLOCK_FAILED: isize = -2;
+const COND_WAIT_FAILED: isize = -3;
+const COND_BROADCAST_FAILED: isize = -4;
+const NULL_THREAD_ARG: isize = -5;
```

Comment:
How would that work with a Rust enum, exactly?

Reply:
Please explain how that would work with a Rust enum.

Reply:
Something like:

```rust
enum ThreadReturnVal {
    Success,
    MutexLockFailed,
    MutexUnlockFailed,
    CondWaitFailed,
    CondBroadcastFailed,
    NullThreadArg,
}

impl std::fmt::Display for ThreadReturnVal {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let msg = match self {
            ThreadReturnVal::Success => "success",
            ThreadReturnVal::MutexLockFailed => "pthread_mutex_lock failed",
            ThreadReturnVal::MutexUnlockFailed => "pthread_mutex_unlock failed",
            ThreadReturnVal::CondWaitFailed => "pthread_cond_wait failed",
            ThreadReturnVal::CondBroadcastFailed => "pthread_cond_broadcast failed",
            ThreadReturnVal::NullThreadArg => "null thread argument",
        };
        write!(f, "{}", msg)
    }
}

impl TryFrom<u32> for ThreadReturnVal {
    type Error = &'static str;
    fn try_from(v: u32) -> Result<Self, Self::Error> {
        match v {
            x if x == ThreadReturnVal::Success as u32 => Ok(ThreadReturnVal::Success),
            x if x == ThreadReturnVal::MutexLockFailed as u32 => Ok(ThreadReturnVal::MutexLockFailed),
            x if x == ThreadReturnVal::MutexUnlockFailed as u32 => Ok(ThreadReturnVal::MutexUnlockFailed),
            x if x == ThreadReturnVal::CondWaitFailed as u32 => Ok(ThreadReturnVal::CondWaitFailed),
            x if x == ThreadReturnVal::CondBroadcastFailed as u32 => Ok(ThreadReturnVal::CondBroadcastFailed),
            x if x == ThreadReturnVal::NullThreadArg as u32 => Ok(ThreadReturnVal::NullThreadArg),
            _ => Err("Unknown enum variant"),
        }
    }
}

// Check for error values returned from pthread test functions
fn check_pthread_error(rv: ThreadReturnVal) -> Result<(), String> {
    match rv {
        ThreadReturnVal::Success => Ok(()),
        _ => Err(rv.to_string())
    }
}
```
```rust
return ThreadReturnVal::NullThreadArg as u32 as *mut libc::c_void;
```
```rust
check_pthread_error(ThreadReturnVal::try_from(rv as u32)?)?;
```

The `TryFrom` is a little more verbose than I thought it would be, but I do think it's a bit clearer.

Reply:
That seems like a long way to go vs. six constants...